tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py

```diff
@@ -115,6 +118,283 @@ def test_serialize(self):
+        args, kwargs = optimizer.update.call_args
+        assert len(args) == 2
+        assert len(kwargs) == 0
+        loss, v1 = args
+
+        # Check the input device.
+        assert backend.get_device_from_array(v1) == input_device
+
+
+class DummyDevice(backend.Device):
+
+    xp = numpy
+    supported_array_types = (numpy.ndarray,)
+
+    def __init__(self, index):
+        self.index = index
+
+    def __eq__(self, other):
+        return isinstance(other, DummyDevice) and other.index == self.index
+
+    # TODO(niboshi): Define name property instead (#7149).
+    def __str__(self):
+        return '@dummy:{}'.format(self.index)
+
+    def send_array(self, array):
+        return array.copy()
+
+
+@testing.parameterize(*testing.product({
+    'omit_device': [True, False],
+    'omit_input_device': [True, False],
+}))
+class TestStandardUpdaterDeviceArgumentFallback(unittest.TestCase):
+    """Tests the fallback behavior regarding device and input_device
+    arguments."""
+
+    def test_device_argument_fallback(self):
+        self.check_device_argument_fallback(
+            initial_model_device=DummyDevice(0),
+            initial_input_device=DummyDevice(1),
+            device_arg=DummyDevice(3),
+            input_device_arg=DummyDevice(4))
+
+    @attr.multi_gpu(2)
+    def test_gpu_to_gpu_transfer(self):
+        initial_model_device = backend.GpuDevice.from_device_id(0)
+        initial_input_device = backend.GpuDevice.from_device_id(0)
+        # GpuDevice is given as device/input_device arguments:
+        # - model GPU-to-GPU transfer should be skipped.
+        # - input GPU-to-GPU transfer should NOT be skipped.
+
+        # device      : GPU 1
+        # input_device: Other device
+        self.check_device_argument_fallback(
+            initial_model_device=initial_model_device,
+            initial_input_device=initial_input_device,
+            device_arg=backend.GpuDevice.from_device_id(1),
+            input_device_arg=DummyDevice(0))
+
+        # device      : GPU 1
+        # input_device: omitted
+        self.check_device_argument_fallback(
+            initial_model_device=initial_model_device,
+            initial_input_device=initial_input_device,
+            device_arg=backend.GpuDevice.from_device_id(1),
+            input_device_arg=None)
+
+        # device      : Other device
+        # input_device: GPU 1
+        self.check_device_argument_fallback(
+            initial_model_device=initial_model_device,
+            initial_input_device=initial_input_device,
+            device_arg=DummyDevice(0),
+            input_device_arg=backend.GpuDevice.from_device_id(1))
+
+        # device      : omitted
+        # input_device: GPU 1
+        self.check_device_argument_fallback(
+            initial_model_device=initial_model_device,
+            initial_input_device=initial_input_device,
+            device_arg=None,
+            input_device_arg=backend.GpuDevice.from_device_id(1))
```

Comment:
As these checks may fail independently, how about separating them?

Reply:
These tests are closely related.
Separating them would make that point less apparent.
I wanted to write this test with parameterization but it's not possible so I wrote in this manner.